src/FuseGPUThreadLoops.cpp

```diff
@@ -284,12 +279,28 @@ class ExtractSharedAllocations : public IRMutator {
             IRMutator::visit(op);
             in_threads = old;
         } else {
-            Interval min_bounds = bounds_of_expr_in_scope(op->min, scope);
-            Interval extent_bounds = bounds_of_expr_in_scope(op->extent, scope);
-            Interval bounds(min_bounds.min, min_bounds.max + extent_bounds.max - 1);
-            scope.push(op->name, bounds);
-            IRMutator::visit(op);
-            scope.pop(op->name);
+            vector<SharedAllocation> old;
+            old.swap(allocations);
+
+            Stmt body = mutate(op->body);
+
+            // Expand any new shared allocations found in the body using the loop bounds.
+            Scope<Interval> scope;
+            scope.push(op->name, Interval(Variable::make(Int(32), op->name + ".loop_min"),
+                                          Variable::make(Int(32), op->name + ".loop_max")));
+            for (SharedAllocation &s : allocations) {
+                if (expr_uses_var(s.size, op->name)) {
+                    s.size = bounds_of_expr_in_scope(s.size, scope).max;
+                }
+            }
+
+            if (!allocations.empty()) {
```

Comment:
It looks like this is being careful to reverse the order of allocations... why is this necessary? I'm concerned because later, the allocations get sorted (by type size).